### ISSUE TITLE
ReactRefreshWebpackPlugin with shared react (module federation & webpack5)

### ISSUE THREAD

**Issue by noreiller** [vercel/next.js#20903]
**What version of Next.js are you using?**

10.0.5

**What version of Node.js are you using?**

14.15.3

**What browser are you using?**

firefox

**What operating system are you using?**

ubuntu

**How are you deploying your application?**

next start

**Describe the Bug**

When I use `webpack@5` and I share react in the `ModuleFederationPlugin`, there is an error in development environment in the ReactRefreshWebpackPlugin. There is no issue in production environment.

```
Uncaught TypeError: originalFactory is undefined
```

The cause is in this generated code:
```js
/******/ 	!function() {
/******/ 		__webpack_require__.i.push(function(options) {
/******/ 			var originalFactory = options.factory;
/******/ 			options.factory = function(moduleObject, moduleExports, webpackRequire) {
/******/ 				var hasRefresh = typeof self !== "undefined" && !!self.$RefreshInterceptModuleExecution$;
/******/ 				var cleanup = hasRefresh ? self.$RefreshInterceptModuleExecution$(moduleObject.id) : function() {};
/******/ 				try {
/******/ 					originalFactory.call(this, moduleObject, moduleExports, webpackRequire);
//                                              ^^^^^ Uncaught TypeError: originalFactory is undefined
/******/ 				} finally {
/******/ 					cleanup();
/******/ 				}
/******/ 			}
/******/ 		})
/******/ 	}();
```

**Expected Behavior**

The `react` module should be loaded on client side, the page should appear and the HMR should be effective.

**To Reproduce**

```bash
yarn create next-app my-app
cd my-app
yarn add @module-federation/nextjs-mf
```

Set the resolution for `webpack@5` in `package.json`:
```json
"resolutions": {
  "webpack": "5.11.1"
}
```
Create the `next.config.js` file:
```js
const { withModuleFederation } = require("@module-federation/nextjs-mf");
const { dependencies } = require("./package.json");

module.exports = {
  webpack: (config, options) => {
    if (options.isServer) {
      return config;
    }

    const newConfig = { ...config };
    const mfConf = {
      mergeRuntime: true,
      name: "nextjs",
      library: { type: config.output.libraryTarget, name: "nextjs" },
      filename: "static/runtime/remoteEntry.js",
      remotes: {},
      shared: {
        react: {
          requiredVersion: dependencies.react,
          eager: true,
        },
      },
    };

    withModuleFederation(newConfig, options, mfConf);

    // This plugin erases the previous externals, so we restore them.
    // Moreover I want to test the share of react, so no react as external.
    newConfig.externals = config.externals;

    return newConfig;
  },
};
```

Run the app
```bash
yarn dev
```

Open the index page on http://localhost:3000

**Comment by arne-sag** [comment]
I'm also facing the same issue. Did you get any solution?

**Comment by noreiller** [comment]
> I'm also facing the same issue. Did you get any solution?

@arne-sag No, I didn't find any solution. I should retry with the latest release since @sokra joined the team and made some webpack changes.

**Comment by ScriptedAlchemy** [comment]
no, you cannot share anything with next.js - especially any deps it internally uses. 
Ive managed a workaround, though needs improvement - https://github.com/module-federation/module-federation-examples/tree/master/nextjs

Next is missing an async `bootstrap` mechanism which allows webpack to figure out if it has React already or needs to download it before initializing. 

Next also uses entrypoint routing, so each entrypoint would need some async boundary (ideally there's only one entrypoint and the router use dynamic imports to route to the right  page, not some entrypoint loader 

We are likely turning to https://github.com/jacob-ebey/mwap in the near future since it supports MF :/ 

For the next few months, I may be able to improve Module Federation on nextjs but it's not likely to be pretty. 

@sokra is working there now, so Vercel do literally own the whole toolchain and can / could solve this

**Comment by alcal1994** [comment]
I get an "Error: Failed to load" when trying to load in a remote component between two next.js apps. Anyone know what might cause this?

I also can't use MergeRuntime as I get an undefined error. 

I can get my remote app working, but fails on host app. I'm using next: "^10.2.0",  and module federation 0.1.1

**Comment by alcal1994** [comment]
This is the next.config from my remote app:
`//module federation configuration
      const { buildId, dev, isServer, defaultLoaders, webpack } = options;
      const mfConf = {
        name: 'form',
        library: { type: config.output.libraryTarget, name: 'form'},
        filename: 'static/runtime/remoteEntry.js',
        remotes:{},
        exposes: {
          './test': './components/layout/test'
        },
        shared:[]
      };

      withModuleFederation(config, options, mfConf);

      if(!isServer){
        config.output.publicPath = 'http://localhost:3000/_next/';
      }

      return config`

This is my next.config from my second app:
`const { buildId, dev, isServer, defaultLoaders, webpack } = options

      const mfConf = {
        name: 'search',
        library: { type: "var", name: "search" },
        filename: 'static/runtime/remoteEntry.js',
        remotes: {
          form: isServer? path.resolve(
            __dirname,
            '../form/.next/server/static/runtime/remoteEntry.js'
          ) 
          : 'form',
        },
        exposes: {},
        shared: [],
      }

      withModuleFederation(config, options, mfConf)

      if(!isServer) {
        config.output.publicPath = "http://localhost:3001/_next/";
      }

      return config`

This is how i'm trying to import it in my host app: 
`const Test = (await import('form/test')).default;`

I have both _document.js files setup with patchSharing, and my host app has the following script tags:
`<script src="http://localhost:3001/_next/static/chunks/webpack.js" />
            <script src="http://localhost:3001/_next/static/runtime/remoteEntry.js" />`

**Comment by ScriptedAlchemy** [comment]
I just updated the plugin and example

**Comment by alcal1994** [comment]
@ScriptedAlchemy Hey Zack, I have my apps setup between my next.js projects. Is there anyway that the host app can change the redux state in the remote app? Does does using redux between the applications work?

**Comment by ScriptedAlchemy** [comment]
yeah it does, just make sure it's a singleton and you'd need to use top-level await to import it because there's no async boundary. 
You could also expose the react store from the host app if you wanted

**Comment by ScriptedAlchemy** [comment]
You can now use the MF plugin on its own. Just make sure react is shared with eager:true

Tested it today, no longer need the nextjs-Mf workaround
